test-utils/src/dev/clickhouse.rs

```diff
@@ -141,6 +106,153 @@ impl ClickHouseInstance {
+            .env("CH_DATASTORE", data_dir.path())
+            .env("CH_TMP_PATH", tmp_path)
+            .env("CH_USER_FILES_PATH", user_files_path)
+            .env("CH_USER_LOCAL_DIR", access_path)
+            .env("CH_FORMAT_SCHEMA_PATH", format_schemas_path)
+            .env("CH_REPLICA_NUMBER", r_number)
+            // There seems to be a bug using ipv6 with a replicated set up
+            // when installing all servers and coordinator nodes on the same
+            // server. For this reason we will be using ipv4 for testing.
+            .env("CH_REPLICA_HOST_01", "127.0.0.1")
+            .env("CH_REPLICA_HOST_02", "127.0.0.1")
+            .env("CH_KEEPER_HOST_01", "127.0.0.1")
+            .env("CH_KEEPER_HOST_02", "127.0.0.1")
+            .env("CH_KEEPER_HOST_03", "127.0.0.1")
+            .spawn()
+            .with_context(|| {
+                format!("failed to spawn `clickhouse` (with args: {:?})", &args)
+            })?;
+
+        let data_path = data_dir.path().to_path_buf();
+        let port: u16 = port.parse()?;
+
+        let result = wait_for_ready(log_path).await;
+        match result {
+            Ok(()) => Ok(Self {
+                data_dir: Some(data_dir),
+                data_path,
+                port,
+                args,
+                child: Some(child),
+            }),
+            Err(e) => Err(e),
+        }
+    }
+
+    /// Start a new ClickHouse keeper on the given IPv6 port.
+    pub async fn new_keeper(
+        port: String,
+        k_id: String,
+        config_path: PathBuf,
+    ) -> Result<Self, anyhow::Error> {
+        // We assume that only 3 keepers will be run, and the ID of the keeper can only
+        // be one of "1", "2" or "3". This is to avoid having to pass the IDs of the
+        // other keepers as part of the function's parameters.
+        if !["1", "2", "3"].contains(&k_id.as_str()) {
+            return Err(ClickHouseError::InvalidKeeperId.into());
+        }
+        // Keepers do not allow a dot in the beginning of the directory, so we must
+        // use a prefix.
+        let data_dir = Builder::new()
+            .prefix("k")
+            .tempdir()
+            .context("failed to create tempdir for ClickHouse Keeper data")?;
+
+        let log_path = data_dir.path().join("clickhouse-keeper.log");
+        let err_log_path = data_dir.path().join("clickhouse-keeper.err.log");
+        let log_storage_path = data_dir.path().join("log");
+        let snapshot_storage_path = data_dir.path().join("snapshots");
+        let args = vec![
+            "keeper".to_string(),
+            "--config-file".to_string(),
+            format!("{}", config_path.display()),
+        ];
+
+        let child = tokio::process::Command::new("clickhouse")
+            .args(&args)
+            .stdin(Stdio::null())
+            .stdout(Stdio::null())
+            .stderr(Stdio::null())
+            .env("CLICKHOUSE_WATCHDOG_ENABLE", "0")
```

Comment:
We should probably do `.env_clear()` here to be safe.

Reply:
Unfortunately that removes the environment variables just before the config files are populated with them so I had to leave it as is.